include/swift/AST/FileUnit.h

```diff
@@ -152,6 +152,20 @@ class FileUnit : public DeclContext {
   /// The order of the results is not guaranteed to be meaningful.
   virtual void getTopLevelDecls(SmallVectorImpl<Decl*> &results) const {}
 
+  /// Finds top-level decls in this file filtered by their attributes.
+  ///
+  /// This does a simple local lookup, not recursively looking through imports.
+  /// The order of the results is not guaranteed to be meaningful.
+  ///
+  /// \param Results Vector collecting the decls.
+  ///
+  /// \param matchAttributes Check on the attributes of a decl to
+  /// filter which decls to fully deserialize. Only decls with accepted
+  /// attributes are deserialized and added to Results.
+  virtual void
+  getTopLevelDeclsWhereAttributesMatch(
+              SmallVectorImpl<Decl*> &Results,
+              llvm::function_ref<bool(DeclAttributes)> matchAttributes) const;
```

Comment:
Why not make it more general like `getTopLevelDeclsWhere(SmallVectorImpl<Decl*> &Results, llvm::function_ref<bool(Decl*)> match)`. We don't get much benefits by limiting it to attribute matches only.

Reply:
The more general function would not work in the deserialization implementation where we want to look at the attributes before deserializing the actual Decl*. Deserializing a Decl deserializes all of it, parents, members, etc. This extra work has been causing issues and we can avoid it by deserializing only the attributes to decide if it is worth a full deserialization which will give use the Decl*.

But I see the need for a more general check. I'm considering extending the matching function for more information available before the full decl: the DeclKind and maybe its name (which would require to serialize it earlier).

Reply:
It makes sense. Thank you for the explanation, Alexis! Could you add some doc-comments here about this reason?

Reply:
Sure! I've added details on how the SerializedASTFile subclass uses it. I kept it simple as to not promise too much.